src/Renderer/LayeredMaterial.js

```diff
@@ -31,31 +22,106 @@ export function unpack1K(color, factor) {
+class LayeredMaterialLayer {
+    constructor(material, options) {
+        this.id = options.id;
+        this.textureOffset = 0; // will be updated in updateUniforms()
+        this.crs = CRS_DEFINES.findIndex(crs => crs.includes(options.tileMT || 'WGS84'));
+        if (this.crs == -1) {
+            console.error('Unkown crs:', options.tileMT);
         }
 
-        fooTexture += 'else return vec4(0.0,0.0,0.0,0.0);}\n';
+        defineLayerProperty(this, 'opacity', options.opacity, 1, (x, y) => (x > 0) != (y > 0));
+        defineLayerProperty(this, 'visible', options.visible, true, () => true);
+        defineLayerProperty(this, 'effect', options.effect, 0);
+        defineLayerProperty(this, 'bias', options.bias, 0);
+        defineLayerProperty(this, 'scale', options.scale, 1);
+        defineLayerProperty(this, 'mode', options.mode, 0);
+        defineLayerProperty(this, 'zmin', options.zmin, 0);
+        defineLayerProperty(this, 'zmax', options.zmax, Infinity);
+
+        this.textures = [];
+        this.offsetScales = [];
+        this.level = EMPTY_TEXTURE_ZOOM;
+        this.needsUpdate = false;
+        this.autoUpdate = true;
+
+        // register this to the material
+        this.material = material;
+        this.material.layers[this.id] = this;
+        this.updateUniforms();
     }
 
-    return fooTexture;
-};
+    setValues(values) {
+        const autoUpdate = this.autoUpdate;
+        this.autoUpdate = false;
+        Object.assign(this, values);
+        this.autoUpdate = autoUpdate;
+        this.updateUniforms();
+    }
 
-// Array not suported in IE
-var fillArray = function fillArray(array, remp) {
-    for (var i = 0; i < array.length; i++)
-        { array[i] = remp; }
-};
+    updateUniforms() {
+        if (this.autoUpdate && this.needsUpdate) {
+            this.material.updateUniforms();
+        }
+    }
 
-var moveElementArray = function moveElementArray(array, oldIndex, newIndex) {
-    array.splice(newIndex, 0, array.splice(oldIndex, 1)[0]);
-};
+    dispose(needsUpdate = true) {
+        // TODO: WARNING  verify if textures to dispose aren't attached with ancestor
+        for (const texture of this.textures) {
+            if (texture instanceof THREE.Texture) {
+                texture.dispose();
+            }
+        }
+        this.level = EMPTY_TEXTURE_ZOOM;
+        this.textures = [];
+        this.offsetScales = [];
+        this.needsUpdate = needsUpdate;
+        this.updateUniforms();
+    }
+
+    setTexture(index, texture, offsetScale) {
+        this.level = (texture && (index == 0)) ? texture.coords.zoom : this.level;
+        this.textures[index] = texture || null;
+        this.offsetScales[index] = offsetScale || identityOffsetScale;
+        this.needsUpdate = true;
+        this.updateUniforms();
+    }
+
+    setTextures(textures, pitchs) {
+        this.dispose(false);
+        const autoUpdate = this.autoUpdate;
+        this.autoUpdate = false;
```

Comment:
`this.autoUpdate` is complex, you could remove it with one update before rendering